File: playwright-csharp/tests/Reporting/EvidenceSupport.cs
using NUnit.Framework;
using Reqnroll.Infrastructure;
using TestTracks.Playwright.CSharp.Specs.Support;

namespace TestTracks.Playwright.CSharp.Specs.Reporting;

public sealed class EvidenceSupport
{
    private readonly ScenarioState _state;
    private readonly ReqnrollOutputHelper _outputHelper;

    public EvidenceSupport(ScenarioState state, ReqnrollOutputHelper outputHelper)
    {
        _state = state;
        _outputHelper = outputHelper;
    }

    public async Task AttachAsync(bool failed)
    {
        Directory.CreateDirectory(_state.Settings.ArtifactsDirectory);

        WriteEvidence("API evidence", _state.ApiEvidence.Text());

        if (_state.Resources.Browser is null)
        {
            return;
        }

        WriteEvidence("Browser", BrowserEvidence());
        WriteEvidenceIfAny("Browser console", _state.Resources.Browser.Evidence.ConsoleMessages);
        WriteEvidenceIfAny("Page errors", _state.Resources.Browser.Evidence.PageErrors);

        if (failed)
        {
            var screenshotPath = Path.Combine(
                _state.Settings.ArtifactsDirectory,
                $"{_state.ScenarioId}-failure.png");

            await _state.Resources.Browser.Page.ScreenshotAsync(new()
            {
                Path = screenshotPath,
                FullPage = true
            });

            _outputHelper.AddAttachment(screenshotPath);
            TestContext.AddTestAttachment(screenshotPath, "Failure screenshot");

            var tracePath = await _state.Resources.Browser.StopTracingAsync(saveTrace: true, _state.ScenarioId);
            if (tracePath is not null)
            {
                _outputHelper.AddAttachment(tracePath);
                TestContext.AddTestAttachment(tracePath, "Playwright trace");
            }
        }
        else
        {
            await _state.Resources.Browser.StopTracingAsync(saveTrace: false, _state.ScenarioId);
        }
    }

    private void WriteEvidence(string title, string content)
    {
        _outputHelper.WriteLine($"== {title} ==");
        _outputHelper.WriteLine(content);
    }

    private void WriteEvidenceIfAny(string title, IReadOnlyList<string> lines)
    {
        if (lines.Count == 0)
        {
            return;
        }

        WriteEvidence(title, string.Join(Environment.NewLine, lines));
    }

    private string BrowserEvidence()
    {
        var browser = _state.Settings.Browser.ToString().ToLowerInvariant();

        return string.Join(
            Environment.NewLine,
            $"Browser: {browser}",
            $"Headless: {_state.Settings.Headless.ToString().ToLowerInvariant()}");
    }
}

File: playwright-java/src/test/java/testtracks/reporting/EvidenceSupport.java
package testtracks.reporting;

import io.cucumber.java.Scenario;
import testtracks.support.ScenarioState;

public final class EvidenceSupport {
  private final ScenarioState state;
  private final Scenario scenario;

  public EvidenceSupport(ScenarioState state, Scenario scenario) {
    this.state = state;
    this.scenario = scenario;
  }

  public void attach(boolean failed) {
    scenario.log("== API evidence ==");
    scenario.log(state.apiEvidence().text());

    var browser = state.resources().browser();
    if (browser == null) {
      return;
    }

    scenario.log("== Browser ==");
    scenario.log(
        "Browser: "
            + state.settings().browser().name().toLowerCase()
            + System.lineSeparator()
            + "Headless: "
            + state.settings().headless());

    logIfAny("Browser console", browser.evidence().consoleMessages());
    logIfAny("Page errors", browser.evidence().pageErrors());

    if (failed) {
      var screenshot =
          browser
              .page()
              .screenshot(new com.microsoft.playwright.Page.ScreenshotOptions().setFullPage(true));
      scenario.attach(screenshot, "image/png", "Failure screenshot");

      var tracePath = browser.stopTracing(true, scenarioArtifactName());
      if (tracePath != null) {
        scenario.log("Playwright trace written to: " + tracePath);
      }
    } else {
      browser.stopTracing(false, state.scenarioId());
    }
  }

  private void logIfAny(String title, java.util.List<String> lines) {
    if (lines.isEmpty()) {
      return;
    }

    scenario.log("== " + title + " ==");
    scenario.log(String.join(System.lineSeparator(), lines));
  }

  private String scenarioArtifactName() {
    var safeName =
        scenario.getName().toLowerCase().replaceAll("[^a-z0-9]+", "-").replaceAll("(^-|-$)", "");
    if (safeName.isBlank()) {
      safeName = "scenario";
    }
    return safeName + "-" + state.scenarioId();
  }
}
